checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,894 @@
 import org.checkerframework.dataflow.cfg.node.Node;
+import org.checkerframework.dataflow.cfg.node.ObjectCreationNode;
+import org.checkerframework.dataflow.expression.JavaExpression;
+import org.checkerframework.dataflow.expression.LocalVariable;
+import org.checkerframework.dataflow.util.NodeUtils;
 import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
 import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.TreePathUtil;
 import org.checkerframework.javacutil.TreeUtils;
+import org.checkerframework.javacutil.TypesUtils;
+import org.plumelib.util.CollectionsPlume;
 
 /**
- * This class contains the Resource Leak Checker's annotation inference algorithm. It contains
- * inference logic for owning annotations on final owning fields. It adds an @Owning annotation on a
- * field if it finds a method that satisfies the @MustCall obligation of the field along some path
- * to the regular exit point.
+ * This class implements the annotation inference algorithm for the Resource Leak Checker. It is
+ * responsible for inferring annotations such as {@code @}{@link Owning} on owning fields and
+ * parameters, @EnsuresCalledMethods on methods, and @InheritableMustCall on class declarations.
+ *
+ * <p>Each instance of this class corresponds to a single control flow graph (CFG), typically
+ * representing a method.
+ *
+ * <p>The algorithm determines if the @MustCall obligation of a field is disposed along some path
+ * leading to the regular exit point of the method. If the obligation is satisfied, it adds
+ * an @Owning annotation on the field and an @EnsuresCalledMethods annotation on the method being
+ * analyzed by this instance. Additionally, if the method being analyzed satisfies the must-call
+ * obligation of all the enclosed owning fields, it adds a @InheritableMustCall annotation on the
+ * enclosing class.
+ *
+ * <p>Note: This class makes the assumption that the must-call set has only one element. Must-call
+ * sets with more than one element may be supported in the future.
+ *
+ * <p>Note: When the -Ainfer flag is used by default, whole-program inference is disabled for the
+ * Resource Leak Checker, and instead, this special mechanism as the best inference mechanism for
+ * the Resource Leak Checker inference is executed. However, for testing and future experimental
+ * purposes, we defined the -AenableWpiForRlc flag to enable whole-program inference (WPI) when
+ * running the Resource Leak Checker inference.
+ *
+ * @see <a
+ *     href="https://checkerframework.org/manual/#resource-leak-checker-inference-algo">Automatic
+ *     Inference of Resource Leak Specifications</a>
  */
 public class MustCallInference {
 
-  /** The set of owning fields. */
-  private final Set<VariableElement> owningFields = new HashSet<>();
+  /** If true, produce debugging diagnostic output. */
+  private static final boolean debug = true;
```

Comment:
@mernst I think you may have added this code.  Do we want this flag set to `false` by default?